src/main/java/org/myrobotlab/service/Cron.java

```diff
@@ -81,29 +85,55 @@ public Task(Cron cron, String id, String cronPattern, String name, String method
 
     @Override
     public void run() {
-      if (cron != null) {
         log.info("{} Cron firing message {}->{}.{}", cron.getName(), name, method, data);
         cron.send(name, method, data);
-      } else {
-        log.error("cron service is null");
-      }
+        cron.history.add(new TaskHistory(id, new Date()));
+        if (cron.history.size() > cron.HISTORY_SIZE) {
+          cron.history.remove(0);
+        }
+        cron.broadcastState();
     }
 
     @Override
     public String toString() {
       return String.format("%s, %s, %s, %s", id, cronPattern, name, method);
     }
   }
+  
+  public static class TaskHistory {
+    public String id;
+    public Date processedTime;
+    
+    public TaskHistory(String id, Date now) {
+      this.id = id;
+      this.processedTime = now;
+    }    
+  }
 
   public final static Logger log = LoggerFactory.getLogger(Cron.class);
 
   private static final long serialVersionUID = 1L;
 
+  /**
+   * history buffer of tasks that have been executed
+   */
+  final protected List<TaskHistory> history = new ArrayList<>();

```

Comment:
Array list is a poor implementation for a FIFO-like queue usage, maybe a linked list would be better. The reason is, removing the head of an array list requires shifting all of the values to the left, an O(n) operation, while removing the head of a linked list involves changing the head pointer and unlinking the head from the second element, an O(1) time operation.

The field type could also be changed to Queue instead of List to better show how it's being used, but we lose precise indexing and can only traverse the collection through iterators or popping the head. Not sure indexing is needed here anyway, but it is a trade-off

Reply:
Yeah, can also get into nasty multi thread bugs with arraylist .. but the problem i was trying to solve is normalized the id.  
I made sure the only area where add & remove are in the same function block ... in theory i should synchronize it or better yet use a queue ... anyway future refactoring ...